From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 4 Feb 2025 20:18:35 +0800
Subject: [PATCH] Leaves: Wool Hopper Counter

Original license: GPL-3.0-only
Original project: https://github.com/LeavesMC/Leaves

This patch is Powered by fabric-carpet(https://github.com/gnembon/fabric-carpet)

diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index f70b842035582b71f4322bae850c157bf4b8ef8e..f7346fab319b2c5b48cf401e28042edbd0e88107 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -199,8 +199,30 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 changed |= action.getAsBoolean();
             }
 
+            // Leaves start - Wool hopper counter
+            if (org.leavesmc.leaves.util.HopperCounter.isEnabled() && org.dreeam.leaf.config.modules.gameplay.WoolHopperCounter.unlimited) {
+                net.minecraft.world.item.DyeColor woolColor = org.leavesmc.leaves.util.WoolUtils.getWoolColorAtPosition(level, entity.getBlockPos().relative(state.getValue(HopperBlock.FACING)));
+                if (woolColor != null) {
+                    for (int i = 0; i < Short.MAX_VALUE; i++) {
+                        changed |= suckInItems(level, entity);
+                        if (!changed) {
+                            break;
+                        } else {
+                            woolHopperCounter(level, pos, state, HopperBlockEntity.getContainerAt(level, pos));
+                        }
+                    }
+                }
+            }
+            // Leaves end - Wool hopper counter
+
             if (changed) {
                 entity.setCooldown(level.spigotConfig.hopperTransfer); // Spigot
+                // Leaves start - Wool hopper counter
+                if (org.leavesmc.leaves.util.HopperCounter.isEnabled() && org.dreeam.leaf.config.modules.gameplay.WoolHopperCounter.unlimited && woolHopperCounter(level, pos, state, HopperBlockEntity.getContainerAt(level, pos))) {
+                    entity.setCooldown(0);
+                    return true;
+                }
+                // Leaves end - Wool hopper counter
                 setChanged(level, pos, state);
                 // Leaves start - Lithium Sleeping Block Entity
                 if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled
@@ -451,6 +473,13 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     // Paper end - Perf: Optimize Hoppers
 
     private static boolean ejectItems(final Level level, final BlockPos blockPos, final HopperBlockEntity self) {
+        // Leaves start - hopper counter
+        if (org.leavesmc.leaves.util.HopperCounter.isEnabled()) {
+            if (woolHopperCounter(level, blockPos, level.getBlockState(blockPos), HopperBlockEntity.getContainerAt(level, blockPos))) {
+                return true;
+            }
+        }
+        // Leaves end - hopper counter
         Container container = org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled ? self.getInsertInventory(level) : getAttachedContainer(level, blockPos, self); // Leaves - Lithium Sleeping Block Entity
         if (container == null) {
             return false;
@@ -472,6 +501,26 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         return hopperPush(level, container, direction, self); // Paper - Perf: Optimize Hoppers
     }
 
+    // Leaves start - hopper counter
+    private static boolean woolHopperCounter(Level level, BlockPos blockPos, BlockState state, @Nullable Container container) {
+        if (container == null) {
+            return false;
+        }
+        net.minecraft.world.item.DyeColor woolColor = org.leavesmc.leaves.util.WoolUtils.getWoolColorAtPosition(level, blockPos.relative(state.getValue(HopperBlock.FACING)));
+        if (woolColor != null) {
+            for (int i = 0; i < container.getContainerSize(); ++i) {
+                if (!container.getItem(i).isEmpty()) {
+                    ItemStack itemstack = container.getItem(i);
+                    org.leavesmc.leaves.util.HopperCounter.getCounter(woolColor).add(level.getServer(), itemstack);
+                    container.setItem(i, ItemStack.EMPTY);
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+    // Leaves end - hopper counter
+
     private static int[] getSlots(final Container container, final Direction direction) {
         if (container instanceof WorldlyContainer worldlyContainer) {
             return worldlyContainer.getSlotsForFace(direction);
@@ -565,6 +614,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     }
 
     public static boolean addItem(final Container container, final ItemEntity entity) {
+        if (org.leavesmc.leaves.util.HopperCounter.isEnabled() && org.dreeam.leaf.config.modules.gameplay.WoolHopperCounter.unlimited && entity.isRemoved()) return false; // Leaves - Wool hopper counter
         boolean changed = false;
         // CraftBukkit start
         if (org.bukkit.event.inventory.InventoryPickupItemEvent.getHandlerList().getRegisteredListeners().length > 0) { // Paper - optimize hoppers
